### ISSUE TITLE
onSuccess might not be called with the right Realm version for async transaction

### ISSUE THREAD

**Issue by beeender** [realm/realm-java#1893]
Now the `onSuccess` relies on the `REALM_CHANGED` triggers the `advance_read` first. If it is running with a unfinished async query, the `REALM_CHANGED` might be swallowed by `updateAsyncQueries` and the Realm doesn't make a `andvance_read`.

In above case, the next message in the handler queue is to run `onSuccess`, but it the function will be called in a wrong earlier Realm version which the changes in the async transaction is not in.

--------------------- Update Jan. 25th 2016 -------------------------
http://stackoverflow.com/questions/34938152/callback-never-be-invoked-when-i-excute-a-asynchronous-transaction-with-realm

See above question on SO. We need to discuss of the user case of `onSuccess`. Currently if the Realm instance is closed, then the `onSuccess` won't be called depends on the timing. But it might not be what user expect.

Shall we just let `onSuccess` be delivered whenever the transaction is done, and document it clearly that in the `onSuccess`, the Realm might not be the latest version?

**Comment by dalinaum** [comment]
@beeender Are you looking into it? Otherwise, I'll start looking into it.

**Comment by cmelchior** [comment]
Good example in that SO post.
I am not sure we are able to deliver the `onSuccess` handler if the Realm gets closed, as if it gets fully closed, we close the Handler and remove all messages. Not to mention the risk of everything getting GC'ed depending on how the user code is.

The pattern is also a bit strange IMO, as putting async behaviour inside a try-with-resources looks like a very error-prone pattern to me (not just with Realm). I can see why it is easy to do though.

IMO, I think we just need to document it clearly. Alternatively, we either log a warning or throw an exception if you close a Realm while a async transaction is still executing. Thoughts, @beeender ?

**Comment by beeender** [comment]
> throw an exception if you close a Realm while a async transaction is still executing 
This would be too strict I think.

And you are right, because of the `HandlerManager` is closed as well, we cannot deliver the `onSuccess` anymore. 

Let's just doc the behaviour for `onSuccess` and closed Realm.

**Comment by cmelchior** [comment]
Would logging a warning make sense? When closing a Realm we could check for any pending async transactions and if there is a callback registered?

**Comment by beeender** [comment]
Yes, I think logging a warning would help. I guess @dalinaum can add the warning when fix this issue.

**Comment by dalinaum** [comment]
I am reading related code, I will try to add logging a warning.

**Comment by zaki50** [comment]
I wrote a test case to reproduce this in #2671.
```
    @Test
    @RunTestInLooperThread
    public void issue2671() {
        final Realm realm = looperThread.realm;

        final RealmResults<Event> all = realm.where(Event.class).findAllAsync();
        all.addChangeListener(new RealmChangeListener<RealmResults<Event>>() {
            @Override
            public void onChange(RealmResults<Event> element) {
                if (!element.isLoaded()) {
                    return;
                }
                Log.e("lh", "asyncQuery(size): " + element.size());
            }
        });

        final String id = UUID.randomUUID().toString();
        Log.e("lh", "before save " + id + "----" + realm.allObjects(Event.class).size());
        realm.executeTransactionAsync(
                new Realm.Transaction() {
                    @Override
                    public void execute(Realm realm1) {
                        Event games = new Event();
                        games.setId(id);
                        games.setName("test-name!!");
                        realm1.copyToRealmOrUpdate(games);
                    }
                }
                , new Realm.Transaction.OnSuccess() {
                    @Override
                    public void onSuccess() {
                        //Not update yet
                        Log.e("lh", "save " + id + " result ...." + realm.allObjects(Event.class).size());
                        realm.refresh();
                        Log.e("lh", "after realm.refresh()-----save " + id + " result ...." + realm.allObjects(Event.class).size());
                        looperThread.testComplete();
                    }
                });
    }
```

```
package io.realm.entities;

import io.realm.RealmObject;
import io.realm.annotations.PrimaryKey;

public class Event extends RealmObject {
    @PrimaryKey
    private String id;
    private String name;

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }
}
```

and got the following result

```
04-30 03:01:54.277  1911  1925 E lh      : before save 667a1d60-9889-4084-b673-1e4e5acce5db----0
04-30 03:01:54.289  1911  1925 E lh      : asyncQuery(size): 0
04-30 03:01:54.299  1911  1925 E lh      : save 667a1d60-9889-4084-b673-1e4e5acce5db result ....0
04-30 03:01:54.299  1911  1925 E lh      : after realm.refresh()-----save 667a1d60-9889-4084-b673-1e4e5acce5db result ....0
```

**Comment by cmelchior** [comment]
This have been fixed by #2780 which was just merged, and will be part of next release (Either 1.0.2 or 1.1.0).
